File: maximumSFRS.ts
/*Maximum Score From Removing Sub String*/
function maximumGain(s: string, x: number, y: number): number {
    const gain = (first: string, second: string, value: number): ((arr: string[]) => number) => {
        return (arr: string[]) => {
            let top = 0, score = 0;
            for (let i = 0; i < arr.length; i++) {
                if (top > 0 && arr[top - 1] === first && arr[i] === second) {
                    top--;
                    score += value;
                } else {
                    arr[top++] = arr[i];
                }
            }
            arr.length = top;
            return score;
        };
    };

    const arr = s.split('');
    const firstPass = x >= y ? gain('a', 'b', x) : gain('b', 'a', y);
    const secondPass = x >= y ? gain('b', 'a', y) : gain('a', 'b', x);

    const score1 = firstPass(arr);
    const score2 = secondPass(arr);

    return score1 + score2;
}

File: MaximumSFRS.py
class Solution: #Maximum Score From Removing Sub Strings
    def maximumGain(self, s: str, x: int, y: int) -> int:
        def remove_pair(s: str, first: str, second: str, points: int) -> tuple[str, int]:
            stack = []
            score = 0
            for c in s:
                if stack and stack[-1] == first and c == second:
                    stack.pop()
                    score += points
                else:
                    stack.append(c)
            return ''.join(stack), score

        # Always remove the higher-value pair first
        if x >= y:
            s, score1 = remove_pair(s, 'a', 'b', x)
            _, score2 = remove_pair(s, 'b', 'a', y)
        else:
            s, score1 = remove_pair(s, 'b', 'a', y)
            _, score2 = remove_pair(s, 'a', 'b', x)

        return score1 + score2
